frontend/src/main/scala/bloop/io/DirChecksum.scala

```
@@ -0,0 +1,106 @@
+   *
+   * @param newChecksum If set, the new `DirChecksum` with the updated
+   *                    `lastModifiedTimes` (if the files have been touched, but
+   *                    not modified).
+   */
+  case class DirUnchanged(newChecksum: Option[DirChecksum]) extends DirStatus
+
+  /** Indicates that the content of the directory has changed. */
+  case object DirChanged extends DirStatus
+
+  /**
+   * Creates a new `DirChecksum`.
+   *
+   * @param dir     The directory to track.
+   * @param pattern The pattern that must be matched by the tracked files.
+   */
+  def apply(dir: AbsolutePath, pattern: String): DirChecksum = {
+    val files = getFiles(dir, pattern)
+    val checksum = filesChecksum(files.map(_._1))
+    new DirChecksum(dir, pattern, files, checksum)
+  }
+
+  /**
+   * Returns all the tracked files inside this directory, associated with their last
+   * modification time.
+   *
+   * @param dir     The directory to track.
+   * @param pattern The pattern to find the files to track.
+   * @return A map associating each tracked file with its last modification time.
+   */
+  private def getFiles(dir: AbsolutePath, pattern: String): List[(AbsolutePath, FileTime)] = {
+    Paths
+      .getAll(dir, pattern)
+      .map { path =>
+        path -> Files.getLastModifiedTime(path.underlying)
+      }
+      .toList
+  }
+
+  /**
+   * Computes the checksum of `files`.
+   *
+   * @param files The files whose checksum must be computed.
+   * @return The checksum.
+   */
+  private def filesChecksum(files: List[AbsolutePath]): Long = {
+    val streams = new Enumeration[InputStream] {
+      val streams = files.map(f => Files.newInputStream(f.underlying)).toIterator
+      override def hasMoreElements(): Boolean = streams.hasNext
+      override def nextElement(): InputStream = streams.next()
+    }
+
+    val checkedStream = new CheckedInputStream(new SequenceInputStream(streams), new Adler32)
+    try {
```

If something here fails, can we report the error in a nice way and assume the worst (e.g. the build has changed)?

---

Done